src/net/tor_address.cpp

```diff
@@ -0,0 +1,195 @@
+
+        std::uint16_t porti = default_port;
+        if (!port.empty())
+        {
+            if (!epee::string_tools::get_xtype_from_string(porti, std::string{port}))
+                return {net::error::kInvalidPort};
+        }
+
+        static_assert(v2_length <= v3_length, "bad internal host size");
+        static_assert(v3_length + sizeof(tld) == sizeof(tor_address::host_), "bad internal host size");
+        return tor_address{host, porti};
+    }
+
+    bool tor_address::_load(epee::serialization::portable_storage& src, epee::serialization::section* hparent)
+    {
+        if (do_serialize<false>(*this, src, hparent))
+        {
+            if (host_[sizeof(host_) - 1] == 0 && (std::strcmp(host_, unknown_host) == 0 || !host_check(host_).has_error()))
+                return true;
+        }
+        static_assert(sizeof(unknown_host) <= sizeof(host_), "bad buffer size");
+        std::memcpy(host_, unknown_host, sizeof(unknown_host)); // include null terminator
+        port_ = 0;
+        return false;
+    }
+
+    bool tor_address::store(epee::serialization::portable_storage& dest, epee::serialization::section* hparent) const
+    {
+        return do_serialize<true>(*this, dest, hparent);
+    }
+
+    tor_address::tor_address(const tor_address& rhs) noexcept
+      : port_(rhs.port_)
+    {
+        std::memcpy(host_, rhs.host_, sizeof(host_));
+    }
+
+    tor_address& tor_address::operator=(const tor_address& rhs) noexcept
+    {
+        if (this != std::addressof(rhs))
+        {
+            port_ = rhs.port_;
+            std::memcpy(host_, rhs.host_, sizeof(host_));
+        }
+        return *this;
+    }
+
+    bool tor_address::is_unknown() const noexcept
+    {
+        static_assert(1 <= sizeof(host_), "host size too small");
+        return host_[0] == '<'; // character is not allowed otherwise
+    }
+
+    bool tor_address::equal(const tor_address& rhs) const noexcept
+    {
+        return port_ == rhs.port_ && is_same_host(rhs);
+    }
+
+    bool tor_address::less(const tor_address& rhs) const noexcept
+    {
+        return std::strcmp(host_str(), rhs.host_str()) < 0 || port() < rhs.port();
+    }
+
+    bool tor_address::is_same_host(const tor_address& rhs) const noexcept
+    {
+        //! \TODO v2 and v3 should be comparable - requires base32
+        return std::strcmp(host_str(), rhs.host_str()) == 0;
+    }
+
+    std::string tor_address::str() const
+    {
+        const std::size_t host_length = std::strlen(host_str());
+        const std::size_t port_length =
+            port_ == 0 ? 0 : std::numeric_limits<std::uint16_t>::digits10 + 2;
+
+        std::string out{};
+        out.reserve(host_length + port_length); 
+        out.assign(host_str(), host_length);
+
+        if (port_ != 0)
+        {
+            out.push_back(':');
+            namespace karma = boost::spirit::karma;
+            karma::generate(std::back_inserter(out), karma::ushort_, port());
```

Comment:
You mean the overhead of reviewing that library itself? I'm _very_ familiar with spirit source for various reasons (provided a couple of small patches), so I feel fairly comfortable using it. I can see why others aren't. FWIW spirit has been put into `contrib/epee/src/http_auth.cpp`. I was thinking of all replacing `boost::lexical_cast` with spirit for performance reasons in epee.

Reply:
> I'm very familiar with spirit source for various reasons (provided a couple of small patches), so I feel fairly comfortable using it.

No worries, if you feel comfortable using it, and more experienced devs don't have a problem with it.

> FWIW spirit has been put into contrib/epee/src/http_auth.cpp

I saw that as well, and it raised my spidey senses. Since it was outside the scope of this PR, I didn't mention it. Vinnie Falco from the Beast project mentioned to me that having Spirit as a dependency adds a lot of review overhead; likely because of the size of the library, and that it deals with serialization (a notoriously hairy domain for security issues). FWIW, that reasoning makes sense to me, and would also serve well as an argument to remove Spirit use from Monero entirely.